### DISCUSSION TITLE
[QUESTION] Alter Filters.user uids list dinamically

### DISCUSSION THREAD

**Discussion by nicholaslopiccolo** [python-telegram-bot/python-telegram-bot#2358] [Q&A]
Hi everyone thanks for this amazing wrapper!!

I'm trying to figure out the creation of a bot that needs to update the filters attached to the handlers to time in time.
I have an admin with a specific UID that have access to the bot (by filtering), this administrator can create new admins or people with a different permission rank that can or cannot use certain features.
Now I created a class that returns a Filters.user() filled with UIDs of who can access the command based on the rank:
```python
def get_by_rank(self, rank):
    wl_users = Filters.user()
    for (key, value) in enumerate(self.ranks):
        if key >= rank:
            for (j, uid) in enumerate(value):
                wl_users.add_chat_ids(uid)
    return wl_users
```
My problem is that when I launch the app those filters are created and attached to the handlers and then to the dispatcher,
now If my first admin creates a new admin this one will have no access to the bot until the restart of the bot there is a way to update this Filter.user() list dynamically?

Example of handlers:
```python
handlers.append(CommandHandler('start', start, white_list.get_by_rank(0)))
handlers.append(CommandHandler('admins_list', admins_list, white_list.get_by_rank(9)))
```
Attachment to the dispatcher:
```python
for handler in handlers:
    dispatcher.add_handler(handler)
```  
Any Ideas?

**Comment by Bibo-Joshi** [comment]
`Filters.user` has an [`add_user_ids`](https://python-telegram-bot.readthedocs.io/en/stable/telegram.ext.filters.html#telegram.ext.filters.Filters.user.add_user_ids) method ;)

**Comment by nicholaslopiccolo** [comment]
> `Filters.user` has an [`add_user_ids`](https://python-telegram-bot.readthedocs.io/en/stable/telegram.ext.filters.html#telegram.ext.filters.Filters.user.add_user_ids) method ;)

This didn't fix the problem but thanks for the hint

**Comment by Bibo-Joshi** [comment]
Uh, sorry, I misread your `get_by_rank` function. Anyway, you'll need to make your instance of `Filters.user`, i.e. the return value of `get_by_rank` available in your callback in order to alter it. You can e.g. store it in `bot/chat/user_data` or make it a constant that you can import if needed. Note that `Filters.user` is not pickable (in case you're using Picklepersistence)

**Comment by nicholaslopiccolo** [comment]
I already perform what you’re talking about, I have a list that stores those values, get_by_rank is inside a class.
I mean is possible to alter the list of users who can access the bot commands through the update of the Filters.
After I use this line:
```python
handlers.append(CommandHandler('start',start,white_list.get_by_rank(0)))
```
Even if I update my data structure with all the new UIDs those new folks can’t access to the bot.
But maybe we’re misunderstandings...

**Comment by Bibo-Joshi** [comment]
What I'm saying is that you need something like 

```python
rank_0_filter = white_list.get_by_rank(0)

handlers.append(CommandHandler('start',start, rank_0_filter))

… 

def add_user_to_rank_0(update, context):
   ...
   rank_0_filter.add_user_ids(…)
```

**Comment by nicholaslopiccolo** [comment]
So I need a pointer to that resource? So If I update the list it’s still in sync? Also Filters.user().add_user_ids accept an array?

**Reply by Bibo-Joshi** [reply]
If you want to edit the `Filters.user` instance, then you need a reference to that instance, yes. If you only update your `self.ranks` lists, your filter won't know about it, so you need to call the instance methods of your filter instance. Ofc, there are other ways to do that, e.g. store a reference to that filter in your custom class and update it from there in addition to updating the ranks.

and yes, `Filters.user.add_user_ids` accepts a list (as per the docs linked above) ;)

**Reply by nicholaslopiccolo** [reply]
Thanks for the quick help, I'll do it right now!!
